File: LeetCode/1116. Print Zero Even Odd/solution.py
# from threading import Semaphore
# class ZeroEvenOdd:
#     def __init__(self, n):
#         self.n = n
#         self.sem0, self.sem1, self.sem2 = Semaphore(1), Semaphore(0), Semaphore(0)
        
        
#     # printNumber(x) outputs "x", where x is an integer.
#     def zero(self, printNumber: 'Callable[[int], None]') -> None:
#         for i in range(1,self.n+1):
#             self.sem0.acquire()
#             printNumber(0)
#             if i&1 == 1:
#                 self.sem1.release()
#             else:
#                 self.sem2.release()

        
        
#     def even(self, printNumber: 'Callable[[int], None]') -> None:
#         for i in range(1,self.n+1):
#             if i&1 == 0:
#                 self.sem2.acquire()
#                 printNumber(i)
#                 self.sem0.release()
        
        
        
#     def odd(self, printNumber: 'Callable[[int], None]') -> None:
#         for i in range(1,self.n+1):
#             if i&1 == 1:
#                 self.sem1.acquire()
#                 printNumber(i)
#                 self.sem0.release()
        
from threading import Condition
import threading

class ZeroEvenOdd:
    def __init__(self, n):
        self.n = n
        self.cv = Condition()
        self.cur = 1
        self.printZero = True

    # printNumber(x) outputs "x", where x is an integer.

    def zero(self, printNumber: 'Callable[[int], None]') -> None:
        while self.cur <= self.n:
            with self.cv:
                while not self.printZero:
                    self.cv.wait()
                if self.cur <= self.n:
                    printNumber(0)
                self.printZero = not self.printZero
                self.cv.notify_all()

    def even(self, printNumber: 'Callable[[int], None]') -> None:
        while self.cur <= self.n:
            with self.cv:
                while self.printZero or self.cur & 1 == 1:
                    self.cv.wait()
                if self.cur <= self.n:
                    printNumber(self.cur)
                    self.cur += 1
                self.printZero = not self.printZero
                self.cv.notify_all()

    def odd(self, printNumber: 'Callable[[int], None]') -> None:
        while self.cur <= self.n:
            with self.cv:
                while self.printZero or self.cur & 1 == 0:
                    self.cv.wait()
                if self.cur <= self.n:
                    printNumber(self.cur)
                    self.cur += 1
                self.printZero = not self.printZero
                self.cv.notify_all()

File: LeetCode/1116. Print Zero Even Odd/solution.cpp
// class Semaphore
// {
// private:
//     int capacity;
//     mutex mtx;
//     condition_variable cv;

// public:
//     Semaphore(int cap)
//         : capacity(cap)
//     {
//     }
//     void aquire()
//     {
//         unique_lock<mutex> lk(mtx);
//         cv.wait(lk, [this]() {
//             return capacity > 0;
//         });
//         capacity--;
//     }
//     void release()
//     {
//         lock_guard<mutex> lk(mtx);
//         capacity++;
//         cv.notify_all();
//     }
// };

// class ZeroEvenOdd
// {
// private:
//     int n;
//     Semaphore sem_zero, sem_even, sem_odd;
//     bool flag;

// public:
//     ZeroEvenOdd(int n)
//         : sem_zero(1), sem_even(0), sem_odd(0)
//     {
//         this->n = n;
//         flag = true;
//     }

//     // printNumber(x) outputs "x", where x is an integer.
//     void zero(function<void(int)> printNumber)
//     {
//         for (int k = 0; k < n; k++)
//         {
//             sem_zero.aquire();
//             printNumber(0);
//             if (flag)
//             {
//                 sem_odd.release();
//                 flag = false;
//             }
//             else
//             {
//                 sem_even.release();
//                 flag = true;
//             }
//         }
//     }

//     void even(function<void(int)> printNumber)
//     {
//         for (int i = 1; i <= n; i++)
//         {
//             // cout << "even: " << i << endl;
//             if (i % 2 == 0){
//                 sem_even.aquire();
                
//                 printNumber(i);
//                 sem_zero.release();
//             }
            
//         }
//     }

//     void odd(function<void(int)> printNumber)
//     {
//         for (int j = 1; j <= n; j++)
//         {
//             // cout << "odd: " << j << endl;
//             if (j % 2 == 1){
//                 sem_odd.aquire();
                
//                 printNumber(j);
//                 sem_zero.release();
//             }
            
//         }
//     }
// };

class Semaphore
{
private:
    int capacity;
    mutex mtx;
    condition_variable cv;

public:
    Semaphore(int cap)
        : capacity(cap)
    {
    }
    void aquire()
    {
        unique_lock<mutex> lk(mtx);
        cv.wait(lk, [this]() {
            return capacity > 0;
        });
        capacity--;
    }
    void release()
    {
        lock_guard<mutex> lk(mtx);
        capacity++;
        cv.notify_all();
    }
};

class ZeroEvenOdd
{
private:
    int n;
    Semaphore sem_zero, sem_even, sem_odd;
    bool flag;

public:
    ZeroEvenOdd(int n)
        : sem_zero(1), sem_even(0), sem_odd(0)
    {
        this->n = n;
        flag = true;
    }

    // printNumber(x) outputs "x", where x is an integer.
    void zero(function<void(int)> printNumber)
    {
        for (int k = 0; k < n; k++)
        {
            sem_zero.aquire();
            printNumber(0);
            if (flag)
            {
                sem_odd.release();
                flag = false;
            }
            else
            {
                sem_even.release();
                flag = true;
            }
        }
    }

    void even(function<void(int)> printNumber)
    {
        for (int i = 1; i <= n; i++)
        {
            // cout << "even: " << i << endl;
            // if ((i&1) == 0)
            if (!(i&1)){
                sem_even.aquire();
                
                printNumber(i);
                sem_zero.release();
            }
            
        }
    }

    void odd(function<void(int)> printNumber)
    {
        for (int j = 1; j <= n; j++)
        {
            // cout << "odd: " << j << endl;
            // if ((j&1) == 1)
            if (j&1){
                sem_odd.aquire();
                
                printNumber(j);
                sem_zero.release();
            }
            
        }
    }
};














// #include<bits/stdc++.h>
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <functional>
#include <fstream>
using namespace std;


ofstream flog("log");

class Semaphore
{
private:
    int capacity;
    mutex mtx;
    condition_variable cv;

public:
    Semaphore(int cap)
        : capacity(cap)
    {
    }
    void aquire()
    {
        unique_lock<mutex> lk(mtx);
        cv.wait(lk, [this]() {
            return capacity > 0;
        });
        capacity--;
    }
    void release()
    {
        lock_guard<mutex> lk(mtx);
        capacity++;
        cv.notify_all();
    }
};

class ZeroEvenOdd
{
private:
    int n;
    Semaphore sem_zero, sem_even, sem_odd;
    bool flag;

public:
    ZeroEvenOdd(int n)
        : sem_zero(1), sem_even(0), sem_odd(0)
    {
        this->n = n;
        flag = true;
    }

    // printNumber(x) outputs "x", where x is an integer.
    void zero(function<void(int)> printNumber)
    {
        for (int k = 0; k < n; k++)
        {
            sem_zero.aquire();
            printNumber(0);
            if (flag)
            {
                sem_odd.release();
                flag = false;
            }
            else
            {
                sem_even.release();
                flag = true;
            }
        }
    }

    void even(function<void(int)> printNumber)
    {
        for (int i = 1; i <= n; i++)
        {
            // cout << "even: " << i << endl;
            // if ((i&1) == 0)
            if (!(i&1)){
                sem_even.aquire();
                
                printNumber(i);
                sem_zero.release();
            }
            
        }
    }

    void odd(function<void(int)> printNumber)
    {
        for (int j = 1; j <= n; j++)
        {
            // cout << "odd: " << j << endl;
            // if ((j&1) == 1)
            if (j&1){
                sem_odd.aquire();
                
                printNumber(j);
                sem_zero.release();
            }
            
        }
    }
};

int main(){
    cout << "hello\n";

    function<void(int)> printNumber = [](int n) {
        cout << n << endl;
    };
    ZeroEvenOdd zeo(5);
    thread ze(&ZeroEvenOdd::zero, &zeo, printNumber);
    thread ev(&ZeroEvenOdd::even, &zeo, printNumber);
    thread od(&ZeroEvenOdd::odd, &zeo, printNumber);
   
    ze.join();
    ev.join();
    od.join();
    // for (int i = 0; i < 10; i++){
    //     cout << (i&1) << endl;
    // }
}